David physics 2

drichte2 opened the pull request:
I have made a few more somewhat large changes, adding an implicit droplet solver which is now working. It shouldn't change much of what was already there, but only adds all the necessary stuff for subroutine "BE_integration"

drichte2 committed:
Broke out the interpolation portion of the droplet solver into its own subroutine for use both RK2 or BE integration

drichte2 committed:
Merge branch 'gpu-opt' into david_physics_2
Updating my branch from the gpu-opt repository

drichte2 committed:
Rearranged a few more things with interpolation, started editing the BE_integration, beginning with position and velocitiy

drichte2 committed:
The implicit solver is running, but not producing same results as rk2. Making sure that both give the same, correct result now

drichte2 committed:
Implicit solver works. Only using newton; about to add LV_solver

drichte2 committed:
Added the LV_solver and tested it to make sure it works. Implemented it as backup to the Newton

drichte2 committed:
Merge branch 'gpu-opt' into david_physics_2
Now that BE solver is working, make sure this is merged properly with gpu-opt branch

drichte2 committed:
Fixed several things up and tested the implicit solver in a couple of different flow settings

sjsprecious commented on src/Makefile:
```diff
@@ -138,7 +138,8 @@ endif
 #-----------------------------------------------------------------------------
 ifeq (${FC},ifort)
-    OPTS         += -g -O3 -xHost -ip -assume byterecl -fp-model precise -ftz -init=snan,arrays -traceback
+#    OPTS         += -g -O3 -xHost -ip -assume byterecl -fp-model precise -ftz -init=snan,arrays -traceback
```
I also realized the `-init=snan,arrays` flag would broke the CM1 run on NCAR's machine using the Intel compiler. I would suggest to removing this line and keeping the updated version only.

sjsprecious replied:
This compiler flag should be kept and the runtime error is fixed by George in the latest `gpu-opt` branch.

sjsprecious commented on src/droplet.F:
```diff
@@ -394,13 +401,13 @@ subroutine droplet_driver(dt,xh,uh,ruh,xf,yh,vh,rvh,yf,zh,mh,rmh,zf,mf,zs,    &
         ELSE
           ! cm1r19:
-          call find_horizontal_location_index (pdata_locind(np,1), x3d, ib, ie+1, xf, ni+2, iflag)
+          call find_horizontal_location_index (pdata_locind(np,1), pdata(np,prx), ib, ie+1, xf, ni+2, iflag)
```
Could @drichte2 explain why using `pdata(np,prx)`, `pdata(np,pry)` and `pdata(np,prz)` instead of using `x3d`, `y3d` and `z3d` here?

drichte2 replied:
This is related to the jflag issue -- at this point, these are identical so it technically doesn't matter which one you use. I switched to "pdata" because these contain all the new values. i.e., rk2_integration and BE_integration end with applying the parcel boundary conditions and then setting pdata(np,prx)=x3d, pdata(np,pry)=y3d, pdata(np,prz)=z3d.

sjsprecious replied:
Thanks @drichte2 . That makes sense to me.

sjsprecious commented on src/solve2.F:
```diff
@@ -2272,7 +2272,7 @@ subroutine solve2(nstep,rbufsz,num_soil_layers,                  &
       ! finish comms for cflquick:
       call mpi_wait(reqc,mpi_status_ignore,ierr)
-      print *,'solve2: cflmax is: ',cflmax
+      !print *,'solve2: cflmax is: ',cflmax
```
could we just delete this line instead of commenting it out?

drichte2 replied:
Sounds good to me. Someone else had put it in there to debug, and it was printing out way too much stuff

drichte2 committed:
Fixed the namelists to include the viscosity and Schmidt numbers

sjsprecious committed:
Remove the problematic intel compiler flags

sjsprecious commented on src/droplet.F:
```diff
@@ -1234,268 +1123,563 @@ subroutine rk2_integration(dt,np,pdata,pdata_locind,iflag,jflag,kflag,xf,xh,uh,r
+      if (flag==1) num1000 = num1000+1
+
+      !In rare cases the nonlinear solvers fail to converge
+      if (isnan(rt_zeros(1)) .OR.  isnan(rt_zeros(2)) .OR. rt_zeros(2)<0) then
```
According to this thread (https://forums.developer.nvidia.com/t/error-isnan-has-not-been-explicitly-declared/132718/6), `isnan` is not a standard extension and may not be supported by some compilers like `PGI`. I would suggest using a better way to detect the NaN value here.

johnmauff replied:
Has the isnan() issue been resolved yet?

sjsprecious replied:
Not yet. I plan to work on this after we have a stable verification framework and namelist.

sjsprecious replied:
Updated by @drichte2

sjsprecious committed:
Add back -init=snan,arrays flag for intel
Use ieee_arithmetic instead of isnan as a standard extension

	modified:   Makefile
	modified:   droplet.F

sjsprecious committed:
Merge branch 'gpu-opt' into david_physics_2

drichte2 committed:
Trying to test the two-way coupling configuration, but turned into testing several other components

drichte2 committed:
Resolved Makefile conflict

drichte2 committed:
Merge branch 'gpu-opt' into david_physics_2
Merging in recent commits to the gpu-opt branch, which hopefully fix the flow problems

sjsprecious committed:
enable -init=snan,arrays for intel compiler

sjsprecious committed:
Merge branch 'gpu-opt' into david_physics_2

Commit:
Cleaning some things up and continuing to test

Commit:
Merge branch 'david_physics_2' of https://github.com/george-bryan/CM1 into david_physics_2

Commit:
Updated the namelist files to have the correct droplet properties

Commit:
After a long search, fixed a small bug where the Gauss-Newton solver could give negative radii. Added a simple check for the guess given by rad_solver2, better than what had been there

Commit:
Turned back on 2-way coupling

Commit:
Fixed conflict with merge

Commit:
Updated all of the baseline namelist files to include droplet settings and comments

sjsprecious committed:
Merge branch 'david_physics_2' of github.com:george-bryan/CM1 into david_physics_2

sjsprecious committed:
Add interface for bcast calls with scalar and array input
Otherwise the Intel compiler will issue an error

	modified:   bcast.F
	modified:   param.F

Commit:
Changed the droplet multiplicity to be 1e6 so the baseline case will work
